[Modes] Add assert that the mode isn't already on the stack

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91573
* __->__ #90770

Redo of #89726 on a clean PR, thanks @voznesenskym for the first draft!

Commit:
[Modes] Add assert that the mode isn't already on the stack

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

samdow commented on c10/core/ModePyObjTrampoline.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace c10 {
+
+// this is basically just a light wrapper around a SafePyObject that also has
+// some way of declaring a push and pop trampoline. This lets us write those
+// methods in the torch namespace where we have access to pybind
```
Basic design of this version looks like:
every time that we push, pop, or copy the stack, we call a python function on the mode that checks the state and updates the boolean on the mode that says whether it's on the stack or not

There's another version of this that @zou3519 and I talked through: the mode stack is actually vector and a set together in a data structure. We only allow push/pop operations on this data structure and it's responsible of keeping track of what's around

Summary of benefits/costs to each
**Pros of C++ data structure (set + vector)**
- easier to understand
- all logic for mode stack lives in C++
  -  no jumps back to Python

**Pros of Python function version (this version)**
- don't have repetition of the same information (vector + set together has code smell, even if there's safety because of the functions we expose)
- user can decide what it means to be on the stack by overriding the `check_mode_state_*` functions
  - right now the default is that two modes of the same Python type can be on the stack but the exact same mode can't be on the stack. Fake Tensor will want to restrict such that no two Fake Tensor modes can exist on the stack at the same time

Any thoughts/preferences welcome!

ezyang replied:
Yeah I feel like this is the big question. Let me offer one more possibility: handling ALL of the state in Python. The reason we moved it to C++ in the first place was because the TLS save load logic only knew how to deal with C++ objects. If we have a way of registering arbitrary Python to be saved and restored, there isn't a reason outside of efficiency to do it in C++

ezyang replied:
Yeah I feel like this is the big question. Let me offer one more possibility: handling ALL of the state in Python. The reason we moved it to C++ in the first place was because the TLS save load logic only knew how to deal with C++ objects. If we have a way of registering arbitrary Python to be saved and restored, there isn't a reason outside of efficiency to do it in C++

samdow commented on torch/utils/_python_dispatch.py:
```diff
@@ -51,6 +53,20 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         _pop_mode()
 
+    def check_mode_state_push(self):
```
bikeshedding on these names welcome

Idea of this is they're the functions that checks that the mode isn't on the stack already and updates its state to reflect a push or pop

samdow commented on c10/core/impl/TorchDispatchModeTLS.cpp:
```diff
@@ -8,22 +7,26 @@ namespace impl {
     c10::impl::tls_set_dispatch_key_included(
         DispatchKey::PythonTLSSnapshot, true);
   }
+  mode->mode_state_push_trampoline();
```
This and L29 are exactly equivalent of what Voz did in the original PR. This just occurs in the C++ so that all push/pop calls are noted (this lets us properly handle, say, reentrant dispatch)

samdow commented on c10/core/impl/TorchDispatchModeTLS.cpp:
```diff
@@ -50,7 +53,15 @@ const TorchDispatchModeTLS& TorchDispatchModeTLS::get_state() {
 }
 
 void TorchDispatchModeTLS::set_state(const TorchDispatchModeTLS& state) {
+  for (const std::shared_ptr<c10::ModePyObjTrampoline>& state :
```
@eellison JFYI, contrary to what I previously said, this change ended up having to be thread safe. So once fake tensor mode is thread safe, the checks on the mode stack should also be thread safe and it should be good to go

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

ezyang commented on c10/core/ModePyObjTrampoline.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace c10 {
+
+// this is basically just a light wrapper around a SafePyObject that also has
```
Is there a reason we don't use PyInterpreter for this?

samdow replied:
Not sure I fully get the question/suggestion, but here's the context. The basis of this was that we needed to call a specific python function (check_mode_state_push/pop) from TorchDispatchModeTLS (which is in c10). The assumption was that c10 didn't want to have a dependency on pybind or any of these things in c10.

So this builds a trampoline that ends up sending a SafePyObject back to the torch namespace. There torch grabs the current PyInterpreter and calls the specific Python function

So the trampoline lets c10 call a function that it knows should be defined on the mode object without making it depend on pybind

ezyang replied:
Yes. What I am saying is, PyInterpreter is precisely this trampoline mechanism. It has a vtable with a big grab bag of things you might want to do to PyObject. You can just add a pair of new virtual methods to it.

samdow replied:
Ah I see--I didn't love the look of getting the interpreter from an object just to have to pass in the object (e.x.     
 `mode->interpreter()->mode_state_push_trampoline(mode)`) but since that's where all the trampolines are, moved it over

ezyang commented on torch/fx/passes/fake_tensor_prop.py:
```diff
@@ -33,6 +34,7 @@ def run_node(self, n: Node):
 
     def propagate(self, *args):
-        with self._mode:
+        ctx_mgr = nullcontext() if hasattr(self._mode, "tracking") and self._mode.tracking.on_stack else self._mode
```
If we are handling reentrant this way it seems worth an actual API for this

ezyang reviewed (approved):
Check my comments, but directionally this all seems fine

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

samdow commented on test/functorch/test_aotdispatch.py:
```diff
@@ -1905,10 +1905,11 @@ def forward(self, x):
         with self.assertRaisesRegex(
-            AssertionError, "Unexpected fake buffer"
+            TypeError,
+            "no implementation found for .*FakeTensor"
```
This is a much harder error for the user but maybe it's also on them for using fake tensor directly

ezyang replied:
ok, I can think of some possibilities for making this error easier, but this is fine. (Probably easiest is to have some logging when we report NotImplemented, and then have the handler say how to turn the logging on.)

ezyang replied:
A more invasive API approach would be to make NotImplemented sigils also sideband report why they were NotImplemented, so we can collect these up and tell the user about it.

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -635,6 +635,9 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
                     assert fake_mode is arg.fake_mode, "Mixing modes NYI"
 
         assert fake_mode is not None
+        if hasattr(fake_mode, "tracking") and fake_mode.tracking.on_stack:
```
could you add a comment explaining this ?

samdow replied:
Added

Commit:
Update on "[Modes] Add assert that the mode isn't already on the stack"


Redo of #89726 on a clean PR, thanks voznesenskym for the first draft!

[ghstack-poisoned]

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
Hi @samdow! This PR breaks the internal release. The error: 
```
TorchDispatchModeTLS.cpp:59:50: error: declaration shadows a local variable [-Werror,-Wshadow]
  for (const std::shared_ptr<c10::SafePyObject>& state : state.stack_) {
```
Here you can find more details: [D42477794](https://www.internalfb.com/diff/D42477794)

samdow commented:
> The error:

> TorchDispatchModeTLS.cpp:59:50: error: declaration shadows a local variable [-Werror,-Wshadow]
  for (const std::shared_ptr<c10::SafePyObject>& state : state.stack_) {

Just put up the reland. It would be great to have these same errors externally since stuff like this isn't breaking internal-only systems, just we aren't using the right flag. IIRC @zou3519 ran into a similar issue with missing compiler flags in https://github.com/pytorch/pytorch/pull/90317
